src/slowlog.c

```diff
@@ -0,0 +1,160 @@
+static uint32_t slowlog_id = 0;
+static inline size_t min(size_t a, size_t b) { return a < b ? a : b; }
+
+
+int slowlog_init(struct slowlog_queue *slowlog)
+{
+    size_t queue_len = 1 + (config.slowlog_max_len - 1) / config.thread;  // round up
+    slowlog->capacity = queue_len;
+    slowlog->entries = cv_calloc(queue_len, sizeof(struct slowlog_entry));
+    slowlog->entry_locks = cv_malloc(queue_len * sizeof(pthread_mutex_t));
+    slowlog->curr = 0;
+    int err;
+    for (size_t i = 0; i != queue_len; i++) {
+        if ((err = pthread_mutex_init(slowlog->entry_locks + i, NULL)) != 0) {
+            LOG(ERROR, "pthread_mutex_init: %s", strerror(err));
+            return CORVUS_ERR;
+        }
+    }
+    return CORVUS_OK;
+}
+
+// Should be called only after all worker threads have stopped
+void slowlog_free(struct slowlog_queue *slowlog)
+{
+    for (size_t i = 0; i != slowlog->capacity; i++) {
+        // no other worker threads will read entry queue any more
+        if (slowlog->entries[i] != NULL) {
+            slowlog_dec_ref(slowlog->entries[i]);
+        }
+        pthread_mutex_destroy(slowlog->entry_locks + i);
+    }
+    cv_free(slowlog->entries);
+    cv_free(slowlog->entry_locks);
+}
+
+struct slowlog_entry *slowlog_create_entry(struct command *cmd, int64_t latency)
+{
+    struct slowlog_entry *entry = cv_calloc(1, sizeof(struct slowlog_entry));
+    entry->id = ATOMIC_INC(slowlog_id, 1);
+    entry->log_time = time(NULL);  // redis also uses time(NULL);
+    entry->latency = latency;
+    entry->refcount = 1;
+
+    const char *bytes_fmt = "(%zd bytes)";
+    char bytes_buf[19];  // strlen("4294967295") + strlen("( bytes)") + 1
+
+    assert(cmd->data.elements > 0);
+    size_t argc = min(cmd->data.elements, SLOWLOG_ENTRY_MAX_ARGC);
+    entry->argc = argc;
+    size_t i = 0;
+
+    // for sub cmd
+    if (cmd->prefix) {
+        const char *c = cmd_extract_prefix(cmd->prefix);
+        assert(c != NULL);
+        entry->argv[i].len = strlen(c);
+        entry->argv[i++].str = (uint8_t*)c;
+        entry->argc = min(argc + 1, SLOWLOG_ENTRY_MAX_ARGC);
+        argc = min(argc, SLOWLOG_ENTRY_MAX_ARGC - 1);
+    }
+    // use the last argument to record total argument counts
+    if (cmd->data.elements > argc) {
+        argc--;
+        const size_t tail_max_len = 64;
+        char tail_buf[tail_max_len];
+        uint8_t *buf = cv_malloc(tail_max_len);
+        int len = snprintf(tail_buf, tail_max_len, "(%zd arguments in total)", cmd->data.elements);
+        len = snprintf((char*)buf, tail_max_len, "$%d\r\n%s\r\n", len, tail_buf);
+        entry->argv[SLOWLOG_ENTRY_MAX_ARGC - 1].str = buf;
+        entry->argv[SLOWLOG_ENTRY_MAX_ARGC - 1].len = len;
+    }
+    size_t max_len = SLOWLOG_ENTRY_MAX_STRING;
+    for (size_t j = 0; j != argc; i++, j++) {
+        struct redis_data *arg = cmd->data.element + j;
+        assert(arg->type == REP_STRING);
+        size_t real_len = mbuf_range_len(arg->buf);
```

Comment:
They are not the same here. `*2\r\n$3\r\nGET\r\n` for example, `mbuf_range_len` calculate the length of `$3\r\nGET\r\n` while `arg->pos.str_len` is only for `GET`.
Here I choose to copy the whole `$3\r\nGET\r\n` instead of just `GET`.